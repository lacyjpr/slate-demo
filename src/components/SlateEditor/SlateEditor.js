import React from 'react';
import { Editor } from 'slate-react';
import { Value } from 'slate';
import SpeechRecognition from 'react-speech-recognition';

const initialValue = Value.fromJSON({
  document: {
    nodes: [
      {
        object: 'block',
        type: 'paragraph',
        nodes: [
          {
            object: 'text',
            leaves: [
              {
                text: 'This text can be edited by the user.',
              },
            ],
          },
        ],
      },
    ],
  },
});

//const transcript = '';

function MarkHotkey(options) {
  const { type, key } = options;

  // Return our "plugin" object, containing the `onKeyDown` handler.
  return {
    onKeyDown(event, change) {
      // Check that the key pressed matches our `key` option.
      if (!event.ctrlKey || event.key != key) return;

      // Prevent the default characters from being inserted.
      event.preventDefault();

      // Toggle the mark `type`.
      change.toggleMark(type);
      return true;
    },
  };
}

const plugins = [
  MarkHotkey({ key: 'b', type: 'bold' }),
  MarkHotkey({ key: 'i', type: 'italic' }),
];

// Define our app...
class SlateEditor extends React.Component {
  // Set the initial value when the app is first constructed.
  state = {
    value: initialValue,
  };

  // On change, update the app's React state with the new editor value.
  onChange = ({ value }) => {
    console.log('onChange called');
    this.setState({ value });
    console.log(value);
    return;
  };

  onTalk = transcript => {
    console.log(transcript);
    console.log('ontalk called');
    console.log(this.editor);
    if (typeof this.editor !== 'undefined' && transcript !== '') {
      console.log(typeof this.editor);
      this.editor.change(change => {
        change.insertText(transcript);
        return;
      });
    }
    // } else if (typeof this.editor !== undefined) {
    //   console.log(typeof this.editor.change);
    //   console.log('editor not undefined');
    //   this.editor.change(change => {
    //     change.extendToStartOfBlock();
    //   });
    //}
  };

  // Render the editor.
  render() {
    const {
      transcript,
      resetTranscript,
      browserSupportsSpeechRecognition,
    } = this.props;

    if (!browserSupportsSpeechRecognition) {
      return null;
    }
    return (
      <div>
        <Editor
          ref={editor => (this.editor = editor)}
          plugins={plugins}
          value={this.state.value}
          onChange={this.onChange}
          renderMark={this.renderMark}
        />
        <div>
          <button onClick={resetTranscript}>Reset</button>
          <span onChange={this.onTalk(transcript)}>{transcript}</span>
        </div>
      </div>
    );
  }

  renderMark = props => {
    switch (props.mark.type) {
      case 'bold':
        return <strong>{props.children}</strong>;
      case 'italic':
        return <em>{props.children}</em>;
    }
  };
}

export default SpeechRecognition(SlateEditor);
