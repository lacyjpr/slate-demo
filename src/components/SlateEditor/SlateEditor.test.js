import React from 'react';
//import renderer from 'react-test-renderer';
import { shallow, mount } from 'enzyme';
import serializer from 'enzyme-to-json/serializer';

import SlateEditor from './SlateEditor';

jest.mock('slate', () => {
  return {
    Editor: () => <div>Heal the world!</div>,
    Raw: {
      deserialize: jest.fn(),
      serialize: jest.fn(),
    },
  };
});

expect.addSnapshotSerializer(serializer);

describe('<SlateEditor />', () => {
  it('renders without crashing', () => {
    shallow(<SlateEditor />);
  });
  // it('renders correctly', () => {
  //   const wrapper = mount(<SlateEditor />);
  //   expect(wrapper).toMatchSnapshot();
  // });
});

// import React from 'react'
// import { mount } from 'enzyme'
// import serializer from 'enzyme-to-json/serializer'
// import App from './App'

// jest.mock('slate', () => {
//     return {
//         Editor: () => (<div>Heal the world!</div>),
//         Raw: {
//             deserialize: jest.fn(),
//             serialize: jest.fn()
//         }
//     }
// })

// // We use the enyme-to-json serializer here
// // Note: If you put usage of serializer into jest.config.json,
// //       it refused to work here.
// expect.addSnapshotSerializer(serializer)

// test('should render correctly', () => {
//     const wrapper = mount(
//         <App />
//     )
//     expect(wrapper).toMatchSnapshot()
// })
