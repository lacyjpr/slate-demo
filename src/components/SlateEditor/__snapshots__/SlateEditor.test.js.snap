// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SlateEditor /> renders correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <SlateEditor />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <Editor
          autoCorrect={true}
          autoFocus={false}
          onChange={[Function]}
          plugins={
                    Array [
                              Object {
                                "onKeyDown": [Function],
                              },
                              Object {
                                "onKeyDown": [Function],
                              },
                            ]
          }
          readOnly={false}
          renderMark={[Function]}
          schema={Object {}}
          spellCheck={true}
          value={
                    Object {
                              "document": Object {
                                "data": Object {},
                                "nodes": Array [
                                  Object {
                                    "data": Object {},
                                    "isVoid": false,
                                    "nodes": Array [
                                      Object {
                                        "leaves": Array [
                                          Object {
                                            "marks": Array [],
                                            "object": "leaf",
                                            "text": "This text can be edited by the user.",
                                          },
                                        ],
                                        "object": "text",
                                      },
                                    ],
                                    "object": "block",
                                    "type": "paragraph",
                                  },
                                ],
                                "object": "document",
                              },
                              "object": "value",
                            }
          }
/>,
        <SpeechRecognitionContainer
          editor={undefined}
/>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "autoCorrect": true,
          "autoFocus": false,
          "onChange": [Function],
          "plugins": Array [
            Object {
              "onKeyDown": [Function],
            },
            Object {
              "onKeyDown": [Function],
            },
          ],
          "readOnly": false,
          "renderMark": [Function],
          "schema": Object {},
          "spellCheck": true,
          "value": Object {
            "document": Object {
              "data": Object {},
              "nodes": Array [
                Object {
                  "data": Object {},
                  "isVoid": false,
                  "nodes": Array [
                    Object {
                      "leaves": Array [
                        Object {
                          "marks": Array [],
                          "object": "leaf",
                          "text": "This text can be edited by the user.",
                        },
                      ],
                      "object": "text",
                    },
                  ],
                  "object": "block",
                  "type": "paragraph",
                },
              ],
              "object": "document",
            },
            "object": "value",
          },
        },
        "ref": [Function],
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "editor": undefined,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <Editor
            autoCorrect={true}
            autoFocus={false}
            onChange={[Function]}
            plugins={
                        Array [
                                    Object {
                                      "onKeyDown": [Function],
                                    },
                                    Object {
                                      "onKeyDown": [Function],
                                    },
                                  ]
            }
            readOnly={false}
            renderMark={[Function]}
            schema={Object {}}
            spellCheck={true}
            value={
                        Object {
                                    "document": Object {
                                      "data": Object {},
                                      "nodes": Array [
                                        Object {
                                          "data": Object {},
                                          "isVoid": false,
                                          "nodes": Array [
                                            Object {
                                              "leaves": Array [
                                                Object {
                                                  "marks": Array [],
                                                  "object": "leaf",
                                                  "text": "This text can be edited by the user.",
                                                },
                                              ],
                                              "object": "text",
                                            },
                                          ],
                                          "object": "block",
                                          "type": "paragraph",
                                        },
                                      ],
                                      "object": "document",
                                    },
                                    "object": "value",
                                  }
            }
/>,
          <SpeechRecognitionContainer
            editor={undefined}
/>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "autoCorrect": true,
            "autoFocus": false,
            "onChange": [Function],
            "plugins": Array [
              Object {
                "onKeyDown": [Function],
              },
              Object {
                "onKeyDown": [Function],
              },
            ],
            "readOnly": false,
            "renderMark": [Function],
            "schema": Object {},
            "spellCheck": true,
            "value": Object {
              "document": Object {
                "data": Object {},
                "nodes": Array [
                  Object {
                    "data": Object {},
                    "isVoid": false,
                    "nodes": Array [
                      Object {
                        "leaves": Array [
                          Object {
                            "marks": Array [],
                            "object": "leaf",
                            "text": "This text can be edited by the user.",
                          },
                        ],
                        "object": "text",
                      },
                    ],
                    "object": "block",
                    "type": "paragraph",
                  },
                ],
                "object": "document",
              },
              "object": "value",
            },
          },
          "ref": [Function],
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "editor": undefined,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
